framework/include/convergence/ParsedConvergence.h

```diff
@@ -0,0 +1,100 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "Convergence.h"
+#include "FunctionParserUtils.h"
+
+class Function;
+
+/**
+ * Evaluates convergence from a parsed expression.
+ */
+class ParsedConvergence : public Convergence, public FunctionParserUtils<false>
+{
+public:
+  static InputParameters validParams();
+
+  ParsedConvergence(const InputParameters & parameters);
+
+  virtual MooseConvergenceStatus checkConvergence(unsigned int iter) override;
+
+  virtual void initialSetup() override;
+
+protected:
+  usingFunctionParserUtilsMembers(false);
+
+  /**
+   * Initializes symbols used in the parsed expression
+   */
+  void initializeSymbols();
+  void initializePostprocessorSymbol(unsigned int i);
+  void initializeFunctionSymbol(unsigned int i);
+  void initializeConvergenceSymbol(unsigned int i);
+  void initializeConstantSymbol(unsigned int i);
+
+  /**
+   * Makes a parsed function
+   *
+   * @param[in] expression   expression to parse
+   */
+  SymFunctionPtr makeParsedFunction(const std::string & expression);
+
+  /**
+   * Updates non-constant symbol values
+   *
+   * @param[in] iter   Iteration index
+   */
+  void updateSymbolValues(unsigned int iter);
+  void updatePostprocessorSymbolValues();
+  void updateFunctionSymbolValues();
+  void updateConvergenceSymbolValues(unsigned int iter);
+
+  /**
+   * Converts a Real value to a bool. Error results if value is not 0 or 1.
+   *
+   * @param[in] value  Real value to convert
+   * @param[in] param  Name of the corresponding input parameter
+   */
+  bool convertRealToBool(Real value, const std::string & param) const;
+
+  FEProblemBase & _fe_problem;
+
+  std::vector<std::string> _symbol_names;
```

Comment:
docstring, especially since values holds names for more than one type of object ?